File: dp-ui/src/api/usageStats.ts
import request from './request'
import type { ApiResponse } from '@/types/api'

// ==================== TypeScript 接口定义 ====================

/** 使用统计记录 */
export interface UsageStats {
  id?: number
  userId?: string
  featureCode?: string
  /** 操作类型: view, query, export, create, edit, delete */
  action?: string
  /** 资源类型: dashboard, report, datasource */
  resourceType?: string
  resourceId?: string
  /** 操作耗时(毫秒) */
  duration?: number
  /** 查询结果行数 */
  resultCount?: number
  /** 错误码(如有) */
  errorCode?: string
  /** 客户端类型: web, mobile, api */
  clientType?: string
  createTime?: string
}

/** 资源访问排行 */
export interface ResourceAccessRank {
  resourceId: string
  resourceType: string
  accessCount: number
  uniqueUsers: number
}

// ==================== DAU / WAU / MAU ====================

/** 获取日活跃用户数 */
export const getDAU = (date: string) => {
  return request.get<ApiResponse<number>>('/usage-stats/dau', { params: { date } })
}

/** 获取周活跃用户数 */
export const getWAU = (weekStart: string) => {
  return request.get<ApiResponse<number>>('/usage-stats/wau', { params: { weekStart } })
}

/** 获取月活跃用户数 */
export const getMAU = (year: number, month: number) => {
  return request.get<ApiResponse<number>>('/usage-stats/mau', { params: { year, month } })
}

// ==================== 功能使用频率 ====================

/** 获取功能使用频率统计 */
export const getFeatureUsage = (startDate: string, endDate: string) => {
  return request.get<ApiResponse<Record<string, number>>>('/usage-stats/feature-usage', {
    params: { startDate, endDate }
  })
}

// ==================== 资源排行 ====================

/** 获取资源访问排行 */
export const getResourceRank = (params: {
  resourceType: string
  topN?: number
  startDate: string
  endDate: string
}) => {
  return request.get<ApiResponse<ResourceAccessRank[]>>('/usage-stats/resource-rank', { params })
}

// ==================== 性能分布 ====================

/** 获取查询性能分布 */
export const getQueryPerformance = (startDate: string, endDate: string) => {
  return request.get<ApiResponse<Record<string, number>>>('/usage-stats/query-performance', {
    params: { startDate, endDate }
  })
}

// ==================== 错误率 ====================

/** 获取错误率统计 */
export const getErrorRate = (startDate: string, endDate: string) => {
  return request.get<ApiResponse<Record<string, number>>>('/usage-stats/error-rate', {
    params: { startDate, endDate }
  })
}

// ==================== CSV 导出 ====================

/** 导出统计数据为 CSV 文件 */
export const exportCSV = (startDate: string, endDate: string, limit?: number) => {
  return request.get<Blob>('/usage-stats/export/csv', {
    params: { startDate, endDate, limit },
    responseType: 'blob'
  })
}

File: dp-core/src/main/java/com/dataplatform/data/service/usage/UsageStats.java
package com.dataplatform.data.service.usage;

import java.time.LocalDateTime;

/**
 * 使用统计实体类
 * 需求: 30.1, 30.2, 30.3, 30.4, 30.5
 */
public class UsageStats {
    private Long id;
    private String userId;
    private String featureCode;       // 功能代码
    private String action;            // 操作类型: view, query, export, create, edit, delete
    private String resourceType;      // 资源类型: dashboard, report, datasource
    private String resourceId;        // 资源ID
    private Long duration;            // 操作耗时(毫秒)
    private Integer resultCount;      // 查询结果行数
    private String errorCode;         // 错误码(如有)
    private String clientType;        // 客户端类型: web, mobile, api
    private String clientIp;
    private LocalDateTime createTime;

    public UsageStats() {}

    // Getters and Setters
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public String getUserId() { return userId; }
    public void setUserId(String userId) { this.userId = userId; }

    public String getFeatureCode() { return featureCode; }
    public void setFeatureCode(String featureCode) { this.featureCode = featureCode; }

    public String getAction() { return action; }
    public void setAction(String action) { this.action = action; }

    public String getResourceType() { return resourceType; }
    public void setResourceType(String resourceType) { this.resourceType = resourceType; }

    public String getResourceId() { return resourceId; }
    public void setResourceId(String resourceId) { this.resourceId = resourceId; }

    public Long getDuration() { return duration; }
    public void setDuration(Long duration) { this.duration = duration; }

    public Integer getResultCount() { return resultCount; }
    public void setResultCount(Integer resultCount) { this.resultCount = resultCount; }

    public String getErrorCode() { return errorCode; }
    public void setErrorCode(String errorCode) { this.errorCode = errorCode; }

    public String getClientType() { return clientType; }
    public void setClientType(String clientType) { this.clientType = clientType; }

    public String getClientIp() { return clientIp; }
    public void setClientIp(String clientIp) { this.clientIp = clientIp; }

    public LocalDateTime getCreateTime() { return createTime; }
    public void setCreateTime(LocalDateTime createTime) { this.createTime = createTime; }
}
